### ISSUE TITLE
External database

### ISSUE THREAD

**Issue by lisiruisusan** [jtamames/SqueezeMeta#658]
Dear author,

I came up with a problem when adding new database to existing projects. What shall I do?

(SqueezeMeta) zzzhou@dell-PowerEdge-R740:~$ add_database.pl /database/users/zzzhou/huichen/Hadza /database/users/zzzhou/phi.dmnd

add_database.pl  - (c) J. Tamames, F. Puente-Sánchez CNB-CSIC, Madrid, SPAIN

This is part of the SqueezeMeta distribution (https://github.com/jtamames/SqueezeMeta)
Please cite: Tamames & Puente-Sanchez, Frontiers in Microbiology 10.3389 (2019). doi: https://doi.org/10.3389/fmicb.2018.03349

 m앤╃󈿗½B0M5l¶g¥whYǿ

Error: Invalid parameter count for option '--db'
Error running command:    /database/users/zzzhou/miniconda3/envs/SqueezeMeta/SqueezeMeta/bin/diamond blastp -q /database/users/zzzhou/huichen/Hadza/results/03.Hadza.faa -p 12 -d  -e 0.001 --id 30 --quiet -b NF -f 6 qseqid qlen sseqid slen pident length evalue bitscore qstart qend sstart send -o /database/users/zzzhou/huichen/Hadza/intermediate/04.Hadza.m앤╃󈿗½B0M5l¶g¥whYǿ

.diamond at /database/users/zzzhou/miniconda3/envs/SqueezeMeta/bin/add_database.pl line 71, <infile1> line 0.000000.
(SqueezeMeta) zzzhou@dell-PowerEdge-R740:~$ 

Yours sincerely,

Li

March 31st 2023

**Comment by jtamames** [comment]
Hello
Please check the manual to find out the correct way of specifying new databases
Best,

**Comment by lisiruisusan** [comment]
Dear author,

Thanks so much for your reply. Problems solved.

Yours sincerely,

Li

March 31st 2023

**Comment by Panda-smile** [comment]
Hello, I would like to ask how to solve this problem? I've encountered it too, but I don't know what's wrong~

db.list：CazyDB /public/home/zhanglj/squeezemetatrans/Squeezemetacazy
/public/home/zhanglj/squeezemetatrans/Squeezemetacazy/CAZyDB.Annotation

run：/public/home/zhanglj/miniconda3/envs/squeezeMeta/SqueezeMeta/utils/add_database.pl shotguntrans db.list

**Comment by jtamames** [comment]
Please follow the instructions in the manual

**Comment by Panda-smile** [comment]
> Please follow the instructions in the manual

sadly, Professor, still can't run successfully，

(squeezeMeta) [zhanglj@admin Squeezemetacazy]$ ls
CAZy_description.txt  CAZy.dmnd
(squeezeMeta) [zhanglj@admin Squeezemetacazy]$ pwd
/public/home/zhanglj/squeezemetatrans/Squeezemetacazy

CazyDB	/public/home/zhanglj/squeezemetatrans/Squeezemetacazy/CAZy	/public/home/zhanglj/squeezemetatrans/Squeezemetacazy/CAZy_description.txt（db.list）

/public/home/zhanglj/miniconda3/envs/squeezeMeta/SqueezeMeta/utils/add_database.pl shotguntrans db.list（command）

Calculating coverage for functions
Can't open EXTDB file CazyDB
Can't open EXTDB file CazyDB

**Comment by jtamames** [comment]
Did you format the database? What is its name?

**Comment by Panda-smile** [comment]
yes i  formatted, its  named  CAZy.dmnd

**Comment by jtamames** [comment]
Can you copy the SqueezeMeta_conf.pl file here?

**Comment by Panda-smile** [comment]
> Can you copy the SqueezeMeta_conf.pl file here?

[SqueezeMeta_conf.pl.txt](https://github.com/jtamames/SqueezeMeta/files/14482472/SqueezeMeta_conf.pl.txt)

**Comment by jtamames** [comment]
Hum... weird. But your issue will probably be solved if you change last line of the SqueezeMeta_conf.pl file:
`$opt_db        = "CazyDB";`
by
`$opt_db        = " /public/home/zhanglj/squeezemetatrans/Squeezemetacazy/db.list"`
and run add_databases.pl again

**Comment by jtamames** [comment]
I guess that at some point you run add_database.pl with CazyDB as database name, and that changed SqueezeMeta_conf.pl

**Comment by Panda-smile** [comment]
still erro~ profeesors


(squeezeMeta) [zhanglj@admin squeezemetatrans]$ cat slurm-118520.out

add_database.pl  - (c) J. Tamames, F. Puente-Sánchez CNB-CSIC, Madrid, SPAIN

This is part of the SqueezeMeta distribution (https://github.com/jtamames/SqueezeMeta)
Please cite: Tamames & Puente-Sanchez, Frontiers in Microbiology 10.3389 (2019). doi: https://doi.org/10.3389/fmicb.2018.03349

 CazyDB(squeezeMeta) [zhanglj@admin squeezemetatrans]$ cat slurm-118520.out

add_database.pl  - (c) J. Tamames, F. Puente-Sánchez CNB-CSIC, Madrid, SPAIN

This is part of the SqueezeMeta distribution (https://github.com/jtamames/SqueezeMeta)
Please cite: Tamames & Puente-Sanchez, Frontiers in Microbiology 10.3389 (2019). doi: https://doi.org/10.3389/fmicb.2018.03349

 CazyDB
  Functional assignment for COGS  Found COG annotation file in /mnt/hpc/home/zhanglj/squeezemetatrans/shotguntrans/results/07.shotguntrans.fun3.cog, skipping
 KEGG  Found KEGG annotation file in /mnt/hpc/home/zhanglj/squeezemetatrans/shotguntrans/results/07.shotguntrans.fun3.kegg, skipping
Can't open EXTDB file public/home/zhanglj/squeezemetatrans/Squeezemetacazy/db.list
 PFAM  Found COG annotation file in /mnt/hpc/home/zhanglj/squeezemetatrans/shotguntrans/results/07.shotguntrans.fun3.cog, skipping

  Calculating coverage for functions
Can't open EXTDB file public/home/zhanglj/squeezemetatrans/Squeezemetacazy/db.list
(squeezeMeta) [zhanglj@admin squeezemetatrans]$ ls
cazybat.slurm      cazysmp.slurm  fastq     shotguntrans      slurm-118507.out  Squeezemetacazy
cazyN3batch.slurm  db.list        hr_data   shotguntransCazy  slurm-118519.out  trans2.txt
cazysbatch.slurm   diamond.slurm  samplist  slurm             slurm-118520.out  分析结果
(squeezeMeta) [zhanglj@admin squeezemetatrans]$ cat  db.list 
CazyDB	/public/home/zhanglj/squeezemetatrans/Squeezemetacazy/cazy	/public/home/zhanglj/squeezemeta(squeezeMeta) [zhanglj@admin squeezemetatrans]$ cd Squeezemetacazy
(squeezeMeta) [zhanglj@admin Squeezemetacazy]$ ls
CAZy_description.txt  cazy.dmnd
(squeezeMeta) [zhanglj@admin Squeezemetacazy]$ pwd
/mnt/hpc/home/zhanglj/squeezemetatrans/Squeezemetacazy
(squeezeMeta) [zhanglj@admin Squeezemetacazy]$

**Comment by jtamames** [comment]
Put the correct path for file db.list in SqueezeMeta_conf.pl

**Comment by Panda-smile** [comment]
Hello Professor, I am not sure if my db.list is correct?
[db.list.txt](https://github.com/jtamames/SqueezeMeta/files/14482984/db.list.txt)

A： CazyDB	/public/home/zhanglj/squeezemetatrans/Squeezemetacazy/cazy	/public/home/zhanglj/squeezemetatrans/Squeezemetacazy/CAZy_description.txt
B： CazyDB	/public/home/zhanglj/squeezemetatrans/Squeezemetacazy	/public/home/zhanglj/squeezemetatrans/Squeezemetacazy/CAZy_description.txt

I put the database and db.list  location as follows：
(squeezeMeta) [zhanglj@admin Squeezemetacazy]$ pwd
/mnt/hpc/home/zhanglj/squeezemetatrans/Squeezemetacazy
(squeezeMeta) [zhanglj@admin Squeezemetacazy]$ ls
CAZy_description.txt  cazy.dmnd  db.list

2. I modified SqueezeMeta_conf.pl 
$opt_db        = "public/home/zhanglj/squeezemetatrans/Squeezemetacazy/db.list";

3. run command nomatter A or B db.list it still erro as bellow：
Can't open EXTDB file db.list
 Functional assignment for COGS  Found COG annotation file in /mnt/hpc/home/zhanglj/squeezemetatrans/shotguntrans/results/07.shotguntrans.fun3.cog, skipping
 KEGG  Found KEGG annotation file in /mnt/hpc/home/zhanglj/squeezemetatrans/shotguntrans/results/07.shotguntrans.fun3.kegg, skipping
Can't open EXTDB file public/home/zhanglj/squeezemetatrans/Squeezemetacazy/db.list
 PFAM  Found COG annotation file in /mnt/hpc/home/zhanglj/squeezemetatrans/shotguntrans/results/07.shotguntrans.fun3.cog, skipping
 Calculating coverage for functions
Can't open EXTDB file public/home/zhanglj/squeezemetatrans/Squeezemetacazy/db.list

**Comment by Panda-smile** [comment]
<img width="641" alt="image" src="https://github.com/jtamames/SqueezeMeta/assets/143200215/e08ef8e5-35f8-4398-bfba-8028cbe6e89b">

![image](https://github.com/jtamames/SqueezeMeta/assets/143200215/96675bf6-c90a-4ec6-acfa-e1b2607b61a2)

**Comment by Panda-smile** [comment]
At the same time, I would like to ask whether I should delete those folders that were generated unsuccessfully in the previous stage, because if the files exist, they will be automatically skipped.

**Comment by Panda-smile** [comment]
Hello Professor, it seems that this annotation is still unsuccessful. Can you please try again to find out what caused the problem?（resluts as below）


(squeezeMeta) [zhanglj@admin squeezemetatrans]$ time /public/home/zhanglj/miniconda3/envs/squeezeMeta/SqueezeMeta/utils/add_database.pl shotguntrans db.list

add_database.pl  - (c) J. Tamames, F. Puente-Sánchez CNB-CSIC, Madrid, SPAIN

This is part of the SqueezeMeta distribution (https://github.com/jtamames/SqueezeMeta)
Please cite: Tamames & Puente-Sanchez, Frontiers in Microbiology 10.3389 (2019). doi: https://doi.org/10.3389/fmicb.2018.03349

  File /mnt/hpc/home/zhanglj/squeezemetatrans/shotguntrans/intermediate/04.shotguntrans.CazyDB.diamond for database CazyDB found, skipping run

  Functional assignment for COGS  Found COG annotation file in /mnt/hpc/home/zhanglj/squeezemetatrans/shotguntrans/results/07.shotguntrans.fun3.cog, skipping
 KEGG  Found KEGG annotation file in /mnt/hpc/home/zhanglj/squeezemetatrans/shotguntrans/results/07.shotguntrans.fun3.kegg, skipping
 CazyDB  Result found in /mnt/hpc/home/zhanglj/squeezemetatrans/shotguntrans/results/07.shotguntrans.fun3.CazyDB for database /mnt/hpc/home/zhanglj/squeezemetatrans/shotguntrans/results/07.shotguntrans.fun3.CazyDB, skipping
 PFAM  Found COG annotation file in /mnt/hpc/home/zhanglj/squeezemetatrans/shotguntrans/results/07.shotguntrans.fun3.cog, skipping

  Calculating coverage for functions
Can't open fun3 CazyDB annotation file /mnt/hpc/home/zhanglj/squeezemetatrans/shotguntrans/results/08.shotguntrans.fun3.CazyDB
  Reading coverage in /mnt/hpc/home/zhanglj/squeezemetatrans/shotguntrans/intermediate/10.shotguntrans.mapcount
  Reading rpkm in /mnt/hpc/home/zhanglj/squeezemetatrans/shotguntrans/intermediate/10.shotguntrans.mapcount
  Now creating cog coverage output in /mnt/hpc/home/zhanglj/squeezemetatrans/shotguntrans/results/12.shotguntrans.cog.funcover
  Now creating kegg coverage output in /mnt/hpc/home/zhanglj/squeezemetatrans/shotguntrans/results/12.shotguntrans.kegg.funcover
  Now creating cog raw reads output in /mnt/hpc/home/zhanglj/squeezemetatrans/shotguntrans/ext_tables/12.shotguntrans.cog.stamp
  Now creating kegg raw reads output in /mnt/hpc/home/zhanglj/squeezemetatrans/shotguntrans/ext_tables/12.shotguntrans.kegg.stamp
  Creating table in /mnt/hpc/home/zhanglj/squeezemetatrans/shotguntrans/results/13.shotguntrans.orftable
  Reading GFF in /mnt/hpc/home/zhanglj/squeezemetatrans/shotguntrans/results/08.shotguntrans.gff
  Reading Diamond hits
  Reading COG list
  Reading KEGG list
  Reading CazyDB list: /public/home/zhanglj/squeezemetatrans/Squeezemetacazy/CAZy_description.txt
  Reading aa sequences
  Reading nt sequences
  Reading rRNA sequences
  Reading tRNA/tmRNA sequences
  Reading ORF information
  Calculating GC content for genes
  Calculating GC content for blastx genes
  Calculating GC content for RNAs
  Reading contig information
  Reading KEGG annotations
  Reading COGs annotations
Can't open fun3 CazyDB annotation file /mnt/hpc/home/zhanglj/squeezemetatrans/shotguntrans/results/08.shotguntrans.fun3.CazyDB
  Reading CazyDB annotations
  Reading Pfam annotations
  Reading RPKMs and Coverages
============
GENE TABLE CREATED: /mnt/hpc/home/zhanglj/squeezemetatrans/shotguntrans/results/13.shotguntrans.orftable
============

  Output in /mnt/hpc/home/zhanglj/squeezemetatrans/shotguntrans/results/21.shotguntrans.stats

real	6m38.079s
user	6m22.469s
sys	0m20.357s

**Comment by jtamames** [comment]
The name of your file seems to be db.list.txt, not db.list

**Comment by Panda-smile** [comment]
In order to upload the file I added the txt suffix

**Comment by Panda-smile** [comment]
> The name of your file seems to be db.list.txt, not db.list

In order to upload the file I added the txt suffix

08.shotguntrans.fun3.CazyDB was not generated in the results folder and I didn't find it either

**Comment by jtamames** [comment]
What happens if you run 07.fun3assign.pl shotguntrans?

**Comment by Panda-smile** [comment]
> 07.fun3assign.pl shotguntrans

07.shotguntrans.fun3.CazyDB已经生成。

![image](https://github.com/jtamames/SqueezeMeta/assets/143200215/fb1af728-9187-4405-bfad-f6d6d8068b29)

**Comment by jtamames** [comment]
Ok. Then just do:
`cp 07.shotguntrans.fun3.CazyDB.txt 08.shotguntrans.fun3.CazyDB.txt`
and try again

**Comment by Panda-smile** [comment]
> 好的。然后只需执行： `cp 07.shotguntrans.fun3.CazyDB.txt 08.shotguntrans.fun3.CazyDB.txt` 在结果目录中 并重试

Professor, it seems to be running, please wait for feedback. （wish successfuly）

![image](https://github.com/jtamames/SqueezeMeta/assets/143200215/c927c4db-5b4b-4be6-9781-0af7a80867aa)

**Comment by Panda-smile** [comment]
I am very grateful to the professor for helping me. I finally succeeded.

**Comment by jtamames** [comment]
Glad to hear. Closing issue

**Comment by Panda-smile** [comment]
> 很高兴听到。结题

1. Hello Professor, I found that there are numerical numbers in the results. These do not correspond to any carbohydrate enzymes. I would like to ask if my annotation is reasonable?

 2. eg: (EC 2.4.1.183), in the yellow area in the picture, I only saw the corresponding EC number ((EC 2.4.1.183)) in the CAZy.Annotation file, but it was not shown in the table with annotations to GT5
 
3. 
![image](https://github.com/jtamames/SqueezeMeta/assets/143200215/b36570e2-0c84-4fdf-89a0-f14b6a26afd2)
![image](https://github.com/jtamames/SqueezeMeta/assets/143200215/f3eb2468-dd40-4855-9098-7386fded2adc)
